skbio/stats/distance/_base.py

```diff
@@ -738,6 +740,118 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
 
         return cls(dm, keys_)
 
+    @classonlymethod
+    @experimental(as_of="0.4.2-dev")
+    def from_pairwise_distances(cls, pairwise_distances):
+        """Create ``DistanceMatrix`` from an iterable of ``PairwiseDistances``.
+
+        Constructs a ``DistanceMatrix`` from the complete set of IDs defined by
+        `pairwise_distances`. All "between" pairwise distances must be defined.
+
+        Parameters
+        ----------
+        pairwise_distances : iterable of PairwiseDistances
+            Pairwise distances to merge into a single ``DistanceMatrix``.
+
+        Returns
+        -------
+        DistanceMatrix
+            Distance matrix representing all pairwise distances between the
+            complete set of IDs present in `pairwise_distances`. The IDs in the
+            distance matrix will be in sorted order as there is no ordering
+            defined by `pairwise_distances`.
+
+        Raises
+        ------
+        TypeError
+            If `pairwise_distances` does not contain ``PairwiseDistances``
+            objects.
+        ValueError
+            If there are conflicting distances between a pair of IDs (e.g., two
+            ``PairwiseDistances`` objects define a different distance between
+            the same pair of IDs).
+        ValueError
+            If there is not at least one ID defined across
+            `pairwise_distances`.
+        ValueError
+            If any "between" distances are missing.
+
+        See Also
+        --------
+        skbio.stats.distance.PairwiseDistances
+        skbio.diversity.beta_diversity
+
+        Examples
+        --------
+        Construct a ``DistanceMatrix`` from two ``PairwiseDistances`` objects
+        collectively defining three IDs, `'A'`, `'B'`, and `'C'`:
+
+        >>> from skbio.stats.distance import DistanceMatrix, PairwiseDistances
+        >>> pwdists = [
+        ...     PairwiseDistances({('A', 'B'): 0.3, ('A', 'C'): 0.1}),
+        ...     PairwiseDistances({('B', 'C'): 0.2})
+        ... ]
+        >>> dm = DistanceMatrix.from_pairwise_distances(pwdists)
+        >>> dm
+        3x3 distance matrix
+        IDs:
+        'A', 'B', 'C'
+        Data:
+        [[ 0.   0.3  0.1]
+         [ 0.3  0.   0.2]
+         [ 0.1  0.2  0. ]]
+
+        Note that the IDs are in sorted order.
+
+        """
+        distances = {}
```

Comment:
That's a lot more mem for numpy than expected. Weird. Thanks though!
On May 12, 2016 2:02 PM, "Jose Navas" <notifications@github.com> wrote:

> In skbio/stats/distance/_base.py
> <https://github.com/biocore/scikit-bio/pull/1354#discussion_r63096665>:
>
> > +        ...     PairwiseDistances({('B', 'C'): 0.2})
> > +        ... ]
> > +        >>> dm = DistanceMatrix.from_pairwise_distances(pwdists)
> > +        >>> dm
> > +        3x3 distance matrix
> > +        IDs:
> > +        'A', 'B', 'C'
> > +        Data:
> > +        [[ 0.   0.3  0.1]
> > +         [ 0.3  0.   0.2]
> > +         [ 0.1  0.2  0. ]]
> > +
> > +        Note that the IDs are in sorted order.
> > +
> > +        """
> > +        distances = {}
>
> There is actually a huge difference. I did some tests to actually check
> the performance of the solution presented here and the solution proposed by
> @wasade <https://github.com/wasade> .
>
> chocker.py:
>
>     d = {}
>     for i in range(num_samples):
>         for j in range(i+1, num_samples):
>             print i, j
>             if (i, j) in d or (j, i) in d:
>                 continue
>             else:
>                 d[(i, j)] = 1
>
>     print d
>
> no_chocker_numpy.py
>
>     d = np.zeros([num_samples, num_samples])
>     for i in range(num_samples):
>         for j in range(i+1, num_samples):
>             print i, j
>             if (i, j) in d or (j, i) in d:
>                 continue
>             else:
>                 d[i][j] = 1.0
>
>     print d
>
> I tried with num_samples = 10000 and num_samples = 20000.
> solution memory (GB) time
> chocker 10k ~34GB 4:32
> numpy 10k ~3GB 8:00
> chocker 20k ~137GB 56:00
> numpy 20k ~12GB 39:04
>
> @wasade <https://github.com/wasade> this will explain why your laptop was
> really unhappy when you tried with 25k hahaha
>
> As we can see, as we scale up, we no only use less memory, we are also
> faster. In this case, storing a dense representation is more efficient than
> relying on python dicts.
>
> The memory values above are the ones reported by maxresident memory in
> /usr/bin/time and the time is elapsed time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1354/files/882e4abe48d9c98bf63f0606f4d901dbedfd9a6f#r63096665>
>